src/dialogs/editor.tsx

```diff
@@ -0,0 +1,203 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+
+import { Alert, AlertActionLink } from '@patternfly/react-core/dist/esm/components/Alert';
+import { Button } from '@patternfly/react-core/dist/esm/components/Button';
+import { Modal, ModalVariant } from '@patternfly/react-core/dist/esm/components/Modal';
+import { TextArea } from '@patternfly/react-core/dist/esm/components/TextArea';
+import { Stack } from '@patternfly/react-core/dist/esm/layouts/Stack';
+
+import cockpit from 'cockpit';
+import type { Dialogs, DialogResult } from 'dialogs';
+import { EventEmitter } from 'event';
+
+import "./editor.scss";
+
+const _ = cockpit.gettext;
+
+class EditorState {
+    error: string | null = null; // if there is an error to show
+    modified: boolean = false; // if there are unsaved changes
+    saving: boolean = false; // saving in progress?
+    tag_at_load: string | null = null; // the one we loaded
+    tag_now: string | null = null; // the one on disk
+    content: string = '';
+    writable: boolean = false;
+}
+
+class Editor extends EventEmitter<{ updated(state: EditorState): void }> {
+    file: cockpit.FileHandle<string>;
+    state: EditorState;
+
+    update(updates: Partial<EditorState>) {
+        Object.assign(this.state, updates);
+        this.emit('updated', { ...this.state });
+    }
+
+    modify(content: string) {
+        this.update({ content, modified: true });
+    }
+
+    load_file() {
+        // Can't do this async because we can't get the tag via await
+        this.file.read()
+                .then(((content: string, tag: string) => {
+                    this.update({ content, tag_now: tag, tag_at_load: tag });
+                }) as any /* eslint-disable-line @typescript-eslint/no-explicit-any */)
+                .catch(error => {
+                    this.update({ error: cockpit.message(error) });
+                });
+    }
+
+    constructor(filename: string) {
+        super();
+        this.file = cockpit.file(filename);
+        this.state = new EditorState();
+
+        this.load_file();
+
+        cockpit.spawn(['test', '-w', filename])
+                .then(() => this.update({ writable: true }))
+                .catch(() => this.update({ writable: false }));
+
+        this.file.watch((_content, tag_now) => {
+            this.update({ tag_now });
+        }, { read: false });
+    }
+
+    async save() {
+        if (!this.state.tag_now)
+            return;
+
+        try {
+            this.update({ saving: true });
+            const tag = await this.file.replace(this.state.content, this.state.tag_now);
+            this.update({ saving: false, modified: false, tag_now: tag, tag_at_load: tag });
+        } catch (exc: any /* eslint-disable-line @typescript-eslint/no-explicit-any */) {
+            this.update({ error: cockpit.message(exc), saving: false });
```

Comment:
There's two common error cases here, right? `access-denied`, which cockpit.message() translates to "Not permitted to perform this action.". That's a bit confusing for the case of "You don't have permissions to edit this file. This should already be intercepted by the `test -w` state, so maybe that's just a corner case.

The other is "the file changed underneath your edit", and I don't think we have a matching error code, at least none of the ones that cockpit.message() knows about is appropriate. So this needs some fine tuning.